File: Java/Q235LowestCommonAncestor.java
/*
@b-knd (jingru) on 10 August 2022 09:24:00
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        
        /*
        - stop when one of the node is found, return root
        - stop when need to go separate way (eg. p smaller need to go left, q greater need to go right: return root)
        - go left if both smaller, go right if both greater
        */
        while(root != p && root != q){
            if(p.val < root.val && q.val > root.val || p.val > root.val && q.val < root.val){
                return root;
            } 
            if(p.val < root.val && q.val < root.val){
                root = root.left;
            } else {
                root = root.right;
            }
        }
        return root;
    }
}

//Runtime: 4 ms, faster than 100.00% of Java online submissions for Lowest Common Ancestor of a Binary Search Tree.
//Memory Usage: 42.9 MB, less than 97.81% of Java online submissions for Lowest Common Ancestor of a Binary Search Tree.

File: Python/Q235LowestCommonAncestor.py
# @b-knd (jingru) on 11 August 2022 10:08:00

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        while root != p and root != q:
            if p.val < root.val and q.val > root.val or p.val > root.val and q.val < root.val:
                return root
            if p.val < root.val and q.val < root.val:
                root = root.left
            else:
                root = root.right
        
        return root

#Runtime: 95 ms, faster than 78.79% of Python3 online submissions for Lowest Common Ancestor of a Binary Search Tree.
#Memory Usage: 18.9 MB, less than 22.99% of Python3 online submissions for Lowest Common Ancestor of a Binary Search Tree.
        
        
        
        
